zookeeper/core/component.py

```diff
@@ -69,305 +69,309 @@ class C:
+                raise ValueError(
+                    "Setting component field values directly is prohibited. Use "
+                    "Zookeeper component configuration to set field values."
+                )
+        except AttributeError:
+            pass
+        return fn(instance, name, value)
 
-# Constants which are used internally during component configuration. They are
-# used as placeholders to indicate to a nested sub-component that an ancestor
-# component has a value for a given field name.
-OVERRIDEN_CONF_VALUE = object()
-NON_OVERRIDEN_CONF_VALUE = object()
+    component_cls.__setattr__ = wrapped_fn
 
 
-def set_field_value(instance, name, value):
-    assert not instance.__component_configured__
-    assert name in instance.__component_fields__
-    field = instance.__component_fields__[name]
+def _wrap_delattr(component_cls: Type) -> None:
+    fn = component_cls.__delattr__  # type: ignore
 
-    if value == OVERRIDEN_CONF_VALUE:
-        instance.__component_fields__[name] = InheritedField(
-            annotated_type=field.annotated_type, is_overriden=True
-        )
-    elif value == NON_OVERRIDEN_CONF_VALUE:
-        if isinstance(field, EmptyField):
-            instance.__component_fields__[name] = InheritedField(
-                annotated_type=field.annotated_type, is_overriden=False
-            )
-    else:
-        object.__setattr__(instance, name, value)
-        instance.__component_fields__[name] = NonEmptyField(
-            annotated_type=field.annotated_type, is_overriden=False
-        )
+    @functools.wraps(fn)
+    def wrapped_fn(instance, name):
+        if name in instance.__component_fields__:
+            raise ValueError("Deleting component field values is prohibited.")
+        return fn(instance, name)
 
+    component_cls.__delattr__ = wrapped_fn
 
-####################################
-# Component class method wrappers. #
-####################################
 
+def _wrap_dir(component_cls: Type) -> None:
+    fn = component_cls.__dir__  # type: ignore
 
-def init_wrapper(init_fn):
-    # Components need to be instantiable without arguments, so check that
-    # `init_fn` does not accept any positional arguments without default values.
-    for i, (name, param) in enumerate(inspect.signature(init_fn).parameters.items()):
-        if (
-            i > 0
-            and param.default == inspect.Parameter.empty
-            and param.kind
-            not in [inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD]
-        ):
-            raise ValueError(
-                "The `__init__` method of a component must not accept any "
-                "positional arguments, as the component configuration process "
-                "requires component classes to be instantiable without arguments."
-            )
+    @functools.wraps(fn)
+    def wrapped_fn(instance) -> Set[str]:
+        return set(fn(instance)) | set(instance.__component_fields__.keys())
```

Comment:
I think `__dir__` usually returns  a `list` correct?

```python
        return list(set(fn(instance)) | set(instance.__component_fields__.keys()))
```

Reply:
Ah I think you might be right, though I'll probably just wrap the set because it's nice to get uniqueness.